src/ints/mouse_ps2_bios.cpp

```diff
@@ -0,0 +1,552 @@
+}
+
+static void cmd_reset()
+{
+    cmd_set_defaults();
+    set_type(MouseType::Standard);
+    reset_counters();
+}
+
+static void cmd_set_scaling_21(const bool enable)
+{
+    terminate_unlick_sequence();
+
+    scaling_21 = enable;
+}
+
+bool MOUSEPS2_NotifyMoved(const float x_rel, const float y_rel)
+{
+    delta_x = MOUSE_ClampRelativeMovement(delta_x + x_rel);
+    delta_y = MOUSE_ClampRelativeMovement(delta_y + y_rel);
+
+    return (std::fabs(delta_x) >= 0.5f) || (std::fabs(delta_y) >= 0.5f);
+}
+
+bool MOUSEPS2_NotifyPressedReleased(const MouseButtons12S new_buttons_12S,
+                                    const MouseButtonsAll new_buttons_all)
+{
+    const auto buttons_old = buttons;
+
+    buttons_12S = new_buttons_12S;
+    buttons_all = new_buttons_all;
+    MOUSEPS2_UpdateButtonSquish();
+
+    return (buttons_old.data != buttons.data);
+}
+
+bool MOUSEPS2_NotifyWheel(const int16_t w_rel)
+{
+#ifdef ENABLE_EXPLORER_MOUSE
+    if (type != MouseType::IntelliMouse && type != MouseType::Explorer)
+        return false;
+#else
+    if (type != MouseType::IntelliMouse)
+        return false;
+#endif
+
+    wheel_counter = static_cast<int8_t>(std::clamp(static_cast<int32_t>(w_rel + wheel_counter),
+                                                   static_cast<int32_t>(INT8_MIN),
+                                                   static_cast<int32_t>(INT8_MAX)));
+    return true;
+}
+
+// ***************************************************************************
+// BIOS interface implementation
+// ***************************************************************************
+
+// TODO: Once the the physical PS/2 mouse is implemented, BIOS has to be changed
+// to interact with I/O ports, not to call PS/2 hardware implementation routines
+// directly (no Cmd* calls should be present in BIOS) - otherwise the
+// complicated Windows 3.x mouse/keyboard support will get confused. See:
+// https://www.os2museum.com/wp/jumpy-ps2-mouse-in-enhanced-mode-windows-3-x/
+// Other solution might be to put interrupt lines low ion BIOS implementation,
+// like this is done in DOSBox X.
+
+static bool packet_4bytes = false;
+
+static bool callback_init    = false;
+static uint16_t callback_seg = 0;
+static uint16_t callback_ofs = 0;
+static RealPt ps2_callback   = 0;
+
+void MOUSEBIOS_Reset()
+{
+    cmd_reset();
+    PIC_SetIRQMask(12, false); // lower IRQ line
+    MOUSEVMM_Deactivate();     // VBADOS seems to expect this
+}
+
+void MOUSEBIOS_SetCallback(const uint16_t pseg, const uint16_t pofs)
+{
+    if ((pseg == 0) && (pofs == 0)) {
+        callback_init = false;
+    } else {
+        callback_init = true;
+        callback_seg  = pseg;
+        callback_ofs  = pofs;
+    }
+}
+
+bool MOUSEBIOS_SetPacketSize(const uint8_t packet_size)
+{
+    if (packet_size == 3)
+        packet_4bytes = false;
+    else if (packet_size == 4)
+        packet_4bytes = true;
+    else
+        return false; // unsupported packet size
+
+    return true;
+}
```

Comment:
If something passes in, say `packet_size = 255`, this function returns false, and `packet_4bytes` also remains false, but then the code paths involving `packet_4bytes` always take the path that involves three packets.

``` c++
Bitu MOUSEBIOS_DoCallback()
{
    if (!packet_4bytes) {
        CPU_Push16(packet[0]);
        CPU_Push16(packet[1]);
        CPU_Push16(packet[2]);
    } else {
        CPU_Push16(static_cast<uint16_t>((packet[0] + packet[1] * 0x100)));
        CPU_Push16(packet[2]);
        CPU_Push16(packet[3]);
    }
```

Is that OK / safe? Or do we need extra protection?

Reply:
Hard to tell what is the real life reaction - SeaBIOS only supports 3-byte packets, Ralf Brown Interrupt List does not specify behavior on unsupported value. In my opinion it is unlikely any software will try to set the wrong size - packet size > 3 only makes sense if the software successfully unlocks the relevant mouse protocol extension; since we are not emulate any such mouse, it won't be able to do so.

Looking at SeaBIOS source code (mouse_15c202, mouse_15c203, ...), when receives invalid value it tends not to change the parameter, just return failure - so this is probably the safest way to do.